### PR TITLE
Use the Text Diff to Work Out the Changeset

### LINKED ISSUE #678: Use the Text Diff to Work Out the Changeset
### Summary
The execution service receives textual diffs from the IDE. In order to work out which cached elements need to be invalidated, we need some way of turning this textual diff into a list of _internal_ identifiers.

### Value
We will be able to evict items from the cache based on the changesets being sent by the IDE.

### Specification

- [ ] Take the text diff and work out the positions that it affects.
- [ ] Use this position information to search the _previous_ IR for the positions that have been affected.
- [ ] Use these IR nodes to query the dependency information for the invalidated identifiers (keying being internal IDs).

### Acceptance Criteria & Test Cases

- There is an API for determining the internal identifiers in the IR that have been invalidated by a given change. 
- These identifiers are used to query the data dependency analysis results for the remainder of the entries that should be invalidated.

### PR DESCRIPTION
**Pull request by 4e6**
### Pull Request Description

close #678 

### Important Notes

PR adds the module that computes a list of IR nodes affected by the text edit.

### Checklist
Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/luna/enso/blob/master/doc/style-guides/scala.md), [Java](https://github.com/luna/enso/blob/master/doc/style-guides/java.md), [Rust](https://github.com/luna/enso/blob/master/doc/style-guides/rust.md) or [Haskell](https://github.com/luna/enso/blob/master/doc/style-guides/haskell.md) style guides as appropriate.
- [x] All code has been tested where possible.

### CODE REVIEW

File: engine/runtime/src/main/scala/org/enso/compiler/context/DiffChangeset.scala [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+  /**
+    * Traverses the IR and returns a list of the most specific (the innermost)
+    * IR identifiers affected by the edit by comparing the source locations.
+    *
+    * @param edit the text edit.
+    * @param source the text source.
+    * @param the IR node.
+    * @return the list of IR identifiers affected by the edit.
+    */
+  def compute(edit: TextEdit, source: Rope, ir: IR): Seq[IR.Identifier] = {
```
Comment (iamrecursion) [inline engine/runtime/src/main/scala/org/enso/compiler/context/DiffChangeset.scala:21]: You may want to depend on a more general interface to the source than our specific `Rope`. Could be `CharSequence` or similar. What do you think?

File: engine/runtime/src/main/scala/org/enso/compiler/context/DiffChangeset.scala [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+      toLocation(edit, source),
+      mutable.Queue(ir),
+      Vector.newBuilder[IR.Identifier]
+    )
+  }
+
+  @scala.annotation.tailrec
+  private def compute(
+    edit: Location,
+    queue: mutable.Queue[IR],
```
Comment (iamrecursion) [inline engine/runtime/src/main/scala/org/enso/compiler/context/DiffChangeset.scala:32]: I don't like that this exposes the fact that it uses a queue to the caller. The caller really shouldn't care. I suggest having it just take an `IR`, and then have a _local_ tail recursive `go` method.

File: engine/runtime/src/main/scala/org/enso/compiler/context/DiffChangeset.scala [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+    acc: mutable.Builder[IR.Identifier, Vector[IR.Identifier]]
+  ): Seq[IR.Identifier] = {
+    if (queue.isEmpty) {
+      acc.result()
+    } else {
+      val ir               = queue.dequeue()
+      val affectedChildren = ir.children.filter(intersect(edit, _))
```
Comment (iamrecursion) [inline engine/runtime/src/main/scala/org/enso/compiler/context/DiffChangeset.scala:39]: Maybe `invalidatedChildren` instead? `affectedChildren` is a bit non-specific.

File: engine/runtime/src/main/scala/org/enso/compiler/context/DiffChangeset.scala [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+  }
+
+  /**
+    * Checks if the IR is affected by the edit.
+    *
+    * @param edit location of the edit.
+    * @param ir the IR node.
+    * @return true if the node is affected by the edit.
+    */
+  private def intersect(edit: Location, ir: IR): Boolean = {
```
Comment (iamrecursion) [inline engine/runtime/src/main/scala/org/enso/compiler/context/DiffChangeset.scala:59]: Don't know why these are private! These are useful things that could be externally useful as well.

File: engine/runtime/src/main/scala/org/enso/compiler/context/DiffChangeset.scala (hunk 1 above)
Comment (4e6) [reply engine/runtime/src/main/scala/org/enso/compiler/context/DiffChangeset.scala:21]: After some considerations... I think we can make use of the CharSequence interface :+1:

**Pull request merged**